spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/planning/RewrittenRowLevelCommand.scala

```diff
@@ -0,0 +1,112 @@
+ *  - the row-level command (such as DeleteFromIcebergTable);
+ *  - the scan relation in the rewrite plan that can be either DataSourceV2Relation or
+ *  DataSourceV2ScanRelation depending on whether the planning has already happened;
+ *  - the current rewrite plan.
+ */
+object RewrittenRowLevelCommand {
+  type ReturnType = (RowLevelCommand, LogicalPlan, LogicalPlan)
+
+  def unapply(plan: LogicalPlan): Option[ReturnType] = plan match {
+    case c: RowLevelCommand if c.rewritePlan.nonEmpty =>
+      val rewritePlan = c.rewritePlan.get
+
+      // both ReplaceData and WriteDelta reference a write relation
+      // but the corresponding scan relation should be at the bottom of the write plan
+      // both the write and scan relations will share the same RowLevelOperationTable object
+      // that's why it is safe to use reference equality to find the needed scan relation
+
+      val allowScanDuplication = c match {
+        // group-based updates that rely on the union approach may have multiple identical scans
+        case _: UpdateTable if rewritePlan.isInstanceOf[ReplaceData] => true
+        case _ => false
+      }
+
+      rewritePlan match {
+        case replaceData: ReplaceData =>
+          replaceData.table match {
+            case DataSourceV2Relation(table, _, _, _, _) =>
+              val scanRelation = findScanRelation(table, replaceData.query, allowScanDuplication)
+              scanRelation.map((c, _, replaceData))
+            case _ =>
+              None
+          }
+      }
+
+    case _ =>
+      None
+  }
+
+  private def findScanRelation(
+      table: Table,
+      plan: LogicalPlan,
+      allowScanDuplication: Boolean): Option[LogicalPlan] = {
+
+    val scanRelations = plan.collect {
+      case r: DataSourceV2Relation if r.table eq table => r
+      case r: DataSourceV2ScanRelation if r.relation.table eq table => r
+    }
+
+    // in some cases, the optimizer replaces the v2 scan relation with a local relation
+    // for example, there is no reason to query the table if the condition is always false
+    // that's why it is valid not to find the corresponding v2 scan relation
+
+    scanRelations match {
+      case relations if relations.isEmpty =>
+        None
+
+      case Seq(relation) =>
+        Some(relation)
+
+      case Seq(relation1: DataSourceV2Relation, relation2: DataSourceV2Relation)
+        if allowScanDuplication && (relation1.table eq relation2.table) =>
+        Some(relation1)
+
+      case Seq(relation1: DataSourceV2ScanRelation, relation2: DataSourceV2ScanRelation)
+        if allowScanDuplication && (relation1.scan eq relation2.scan) =>
+        Some(relation1)
+
+      case Seq(relation1, relation2) if allowScanDuplication =>
+        throw new AnalysisException(s"Row-level scan relations don't match: $relation1, $relation2")
+
+      case relations if allowScanDuplication =>
+        throw new AnalysisException(s"Expected 2 row-level scan relations: $relations")
```

Comment:
What is the benefit of failing if you thought there may be 2? What about when users have turned off dynamic filtering or something that would have created the second relation?

Reply:
It is a little hard to explain the logic here without seeing the UPDATE code. I'll remove it for now and then we can discuss it again for UPDATEs.

Reply:
Removed for now.